Framework/TestHelpers/inc/MantidTestHelpers/IndirectFitDataCreationHelper.h

```diff
@@ -0,0 +1,111 @@
+#ifndef MANTID_INDIRECTFITDATACREATIONHELPER_H_
+#define MANTID_INDIRECTFITDATACREATIONHELPER_H_
+
+#include "MantidAPI/AnalysisDataService.h"
+#include "MantidAPI/MatrixWorkspace_fwd.h"
+#include "MantidHistogramData/BinEdges.h"
+
+#include <string>
+
+#include <boost/variant.hpp>
+#include <boost/variant/apply_visitor.hpp>
+#include <boost/variant/static_visitor.hpp>
+
+namespace Mantid {
+namespace IndirectFitDataCreationHelper {
+
+/// Functions used in the creation of workspaces
+Mantid::API::MatrixWorkspace_sptr createWorkspace(int const &numberOfSpectra);
+Mantid::API::MatrixWorkspace_sptr createInstrumentWorkspace(int const &xLength,
+                                                            int const &yLength);
+Mantid::API::MatrixWorkspace_sptr
+setWorkspaceEFixed(Mantid::API::MatrixWorkspace_sptr workspace,
+                   int const &xLength);
+Mantid::API::MatrixWorkspace_sptr
+setWorkspaceBinEdges(Mantid::API::MatrixWorkspace_sptr workspace,
+                     int const &yLength,
+                     Mantid::HistogramData::BinEdges const &binEdges);
+Mantid::API::MatrixWorkspace_sptr
+setWorkspaceBinEdges(Mantid::API::MatrixWorkspace_sptr workspace,
+                     int const &xLength, int const &yLength);
+Mantid::API::MatrixWorkspace_sptr
+setWorkspaceProperties(Mantid::API::MatrixWorkspace_sptr workspace,
+                       int const &xLength, int const &yLength);
+Mantid::API::MatrixWorkspace_sptr
+createWorkspaceWithInstrument(int const &xLength, int const &yLength);
+
+/// Simple struct to set up the ADS with the configuration required
+struct SetUpADSWithWorkspace {
+
+  template <typename T>
+  SetUpADSWithWorkspace(std::string const &inputWSName, T const &workspace) {
+    Mantid::API::AnalysisDataService::Instance().addOrReplace(inputWSName,
+                                                              workspace);
+  }
+
+  template <typename T>
+  void addOrReplace(std::string const &workspaceName, T const &workspace) {
+    Mantid::API::AnalysisDataService::Instance().addOrReplace(workspaceName,
+                                                              workspace);
+  }
+
+  bool doesExist(std::string const &workspaceName) {
+    return Mantid::API::AnalysisDataService::Instance().doesExist(
+        workspaceName);
+  }
+
+  Mantid::API::MatrixWorkspace_sptr
+  retrieveWorkspace(std::string const &workspaceName) {
+    return boost::dynamic_pointer_cast<Mantid::API::MatrixWorkspace>(
+        Mantid::API::AnalysisDataService::Instance().retrieve(workspaceName));
+  }
+
+  ~SetUpADSWithWorkspace() {
+    Mantid::API::AnalysisDataService::Instance().clear();
+  }
+};
+
+/// This struct gives the option of having an ADS instance with no destructor
+struct SetUpADSWithNoDestructor {
```

Comment:
In my opinion, this is a huge bunch of copied code with the only advantage of having a destructor, i.e. clearing the ADS. I would propose to find a better way. Spontaneously, I think of adding something like `clear` to `SetUpADSWithWorkspace` and delete `SetUpADSWithNoDestructor`.
It may be worth to thing more oo oriented (copy class, friend class, inherit class). This may be a second place to look in order to not duplicate code.